modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -3465,52 +3518,30 @@ else if (cntr == maxCntr.cnt)
                 Long histCntr = e0.getValue().partitionHistoryCounters(top.groupId()).get(p);
 
                 if (histCntr != null && maxCntrObj.nodes.contains(e0.getKey())) {
-                    if (histCntr <= minCntr) {
-                        partHistSuppliers.put(e0.getKey(), top.groupId(), p, histCntr);
+                    Long ceilingMinReserved = minCntrs.ceiling(histCntr);
 
-                        haveHistory.add(p);
+                    if (ceilingMinReserved != null) {
+                        partHistSuppliers.put(e0.getKey(), top.groupId(), p, ceilingMinReserved);
```

Comment:
Why not partHistSuppliers.put(e0.getKey(), top.groupId(), p, histCntr); ?
It will work both ways, but it seems more intuitive to distribute actual reserved counter.

Reply:
Comment above.